### DISCUSSION TITLE
Properly avoiding garbage collection of a timer

### DISCUSSION THREAD

**Discussion by tigersoul925** [Hammerspoon/hammerspoon#3434] [Q&A]
Say I need a long timer which I need to make sure isn't garbage collected, and I also don't want to use a global variable because I may add more than one of these timers for whatever purpose in a function, is this really how intricate it has to be? This can't be made simpler?

```
  local tooBrightNotifyTimer
  tooBrightNotifyTimer = hs.timer.doAfter(1000,function()
    doDockActions({"keylight"},force)
    local timer = tooBrightNotifyTimer
    timer = nil
  end)
```

**Comment by cmsj** [comment]
There certainly needs to be some way to hang on to a reference to the timer, to protect it from GC.

I take your point about potentially needing multiple instances of "the same" timer, and I note that timer callbacks currently receive zero arguments...

We could start passing the timer object to the callback. This would let you have a single global table that you insert new timers into, and then when the callbacks determine the timer is finished with, they are able to remove themselves from that table.

Would that fit your needs?

**Reply by tigersoul925** [reply]
Hmm, not 100% how this is reflected in code, but it sounds good.

**Comment by latenitefilms** [comment]
Slightly simpler?

```lua
globalTimers = {}
globalTimers.tooBright = hs.timer.doAfter(1, function()
    doDockActions({"keylight"},force)
    globalTimers.tooBright = nil
end)
```

**Reply by tigersoul925** [reply]
It's simpler. but if this is in a function that gets called to set up a second instance of the timer, one timer will lose its reference and potentially be gc'd. As far as I understand it anyway.
